### PR TITLE
Core: added updated_after filter to list_replicas Closes #3134

### LINKED ISSUE #3134: Add updated_after filtering for list_replicas
Motivation
----------
The idea is to only list files in list_replicas which are created after a certain timestamp. This will help workflows (such as T0) which continuously query for new replicas to quickly assess the files added to a dataset.

### PR DESCRIPTION
**Pull request by lgoossen**
Core: added updated_after filter to list_replicas Closes #3134

### CODE REVIEW

**Comment by lgoossen** [comment]
lots of whitespace complaints by flake8 (seriously ?)
and one missing time import
not sure about all the other failures

File: lib/rucio/api/replica.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 # - Hannes Hansen <hannes.jakob.hansen@cern.ch>, 2019
 # - Andrew Lister <andrew.lister@stfc.ac.uk>, 2019
 # - Martin Barisits <martin.barisits@cern.ch>, 2019
+# - Luc Goossens, <luc.goossens@cern.ch>, 2020
```
Comment (mlassnig) [inline lib/rucio/api/replica.py]: no comma ;-)

Comment (mlassnig) [reply lib/rucio/api/replica.py]: (same for the other files)

File: lib/rucio/core/replica.py [hunk 2]
```diff
@@ -664,40 +665,36 @@ def _list_replicas_for_datasets(dataset_clause, state_clause, rse_clause, sessio
...
-                                   models.RSE.deleted == false(),
-                                   state_clause,
-                                   or_(*replica_condition),
-                                   or_(*rse_clause))
+        filters = [models.RSEFileAssociation.rse_id == models.RSE.id,
+                   models.RSE.deleted == false(),
+                   or_(*replica_condition),
+                   ]
+
+        if state_clause is not None:
```
Comment (mlassnig) [inline lib/rucio/core/replica.py:688]: any reason for the `is not None`? would it suffice to say `if state_clause:` ?

File: lib/rucio/tests/test_replica.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
 from datetime import datetime, timedelta
 from json import dumps, loads
 
+import time
```
Comment (mlassnig) [inline lib/rucio/tests/test_replica.py:31]: sort alphabetically

File: lib/rucio/tests/test_replica.py [hunk 4]
```diff
@@ -27,6 +27,7 @@
 from datetime import datetime, timedelta
```
Comment (mlassnig) [inline lib/rucio/tests/test_replica.py:28]: add author

File: lib/rucio/tests/test_replica.py [hunk 5]
```diff
@@ -501,6 +502,40 @@ def test_set_tombstone(self):
...
+    def test_list_replicas_with_updated_after(self):
+        """ REPLICA (CORE): Add and list file replicas with updated_after filter """
+        scope = InternalScope('mock')
+        root = InternalAccount('root')
+        mock = get_rse_id(rse='MOCK')
+        dsn = 'ds_ua_test_%s' % generate_uuid()
+        add_did(scope=scope, name=dsn, type='DATASET', account=root)
+        #
+        t0 = datetime.utcnow()
+        time.sleep(2)
```
Comment (mlassnig) [inline lib/rucio/tests/test_replica.py:515]: not really happy about the sleeps, but i don't see how this could be done otherwise

**Review by mlassnig — COMMENTED** [review]
looks good, few tiny comments

File: lib/rucio/core/replica.py (hunk 2 above)
Comment (lgoossen) [reply lib/rucio/core/replica.py:688]: no because clauses have no method to cast into boolean

File: lib/rucio/tests/test_replica.py (hunk 3 above)
Comment (lgoossen) [reply lib/rucio/tests/test_replica.py:31]: just tell me where you want it cause according to Martin it's not even in the right section
and of course also white lines matter

**Review by mlassnig — APPROVED** [review]

**Pull request merged**